src/recovery.rs

```diff
@@ -150,15 +148,24 @@ pub fn recover_sealed_memtables(
                 if let Some(handle) = partitions_lock.get(&item.partition) {
                     let tree = &handle.tree;
 
-                    watermarks
-                        .entry(item.partition)
-                        .and_modify(|prev| {
-                            prev.lsn = prev.lsn.max(batch.seqno);
-                        })
-                        .or_insert_with(|| EvictionWatermark {
-                            partition: handle.clone(),
-                            lsn: batch.seqno,
-                        });
+                    match watermarks
+                        .binary_search_by(|watermark| watermark.partition.name.cmp(&item.partition))
+                    {
+                        Ok(index) => {
+                            if let Some(prev) = watermarks.get_mut(index) {
+                                prev.lsn = prev.lsn.max(batch.seqno);
+                            }
+                        }
+                        Err(index) => {
+                            watermarks.insert(
+                                index,
+                                EvictionWatermark {
+                                    partition: handle.clone(),
+                                    lsn: batch.seqno,
+                                },
+                            );
```

Comment:
FWIW the recovery code here is not really a performance hot path, but inserting with index can require shifting a lot of items, so the hash map solution may end up performing better when there are a lot of partitions.
Not to mention the hash map code is more compact.
So I would only change this if there's a proven reason it performs better than the previous solution.